once/_iterator_wrappers.py

```diff
@@ -24,142 +26,229 @@ def __init__(self) -> None:
+    # loop to loop through to the GENERATING so it can complete and unblock the WAITING coroutine.
+    # In the sync case, we also need a sleep(0) for a different reason. While a thread is in the
+    # WAITING state, it would otherwise hog the Global Interpreter Lock in its while loop until
+    # the interprer decides to switch threads. With N threads, N - 1 of them would be in the
+    # WAITING state, and a round-robin interpreter would give each of their while loops the
+    # thread switching time in execution before reaching the single thread in the GENERATING state.
+    # Theoretically, this could result in the GIL only working on the thread in the GENERATING
+    # state 1/Nth of the time, without a time.sleep(0) to manually trigger a GIL switch. In
+    # practice, removing the time.sleep(0) call would result in large drops in performance when
+    # running the multithreaded unit tests.
+    @property
+    @abc.abstractmethod
+    def generating(self) -> bool:
+        raise NotImplementedError()
+
+    @generating.setter
+    @abc.abstractmethod
+    def generating(self, val: bool):
+        raise NotImplementedError()
+
+    def compute_next_action(
+        self, result: IteratorResults, i: int
+    ) -> typing.Tuple[_IteratorAction, typing.Any]:
+        """Must be called with lock."""
+        if i == len(result.items):
+            if result.finished:
+                if result.exception:
+                    raise result.exception
+                return _IteratorAction.RETURNING, None
+            if self.generating:
+                return _IteratorAction.WAITING, None
+            else:
+                # If all of these functions are called with locks, we will never have more than one
+                # caller have GENERATING at any time.
+                self.generating = True
+                return _IteratorAction.GENERATING, None
+        else:
+            return _IteratorAction.YIELDING, result.items[i]
+
+    def record_successful_completion(self, result: IteratorResults):
+        """Must be called with lock."""
+        result.finished = True
+        self.generating = False
+        self.generator = None  # Allow this to be GCed.
+        self.callable = None  # Allow this to be GCed.
+
+    def record_item(self, result: IteratorResults, item: typing.Any):
+        self.generating = False
+        result.items.append(item)
+
+    def record_exception(self, result: IteratorResults, exception: Exception):
+        """Must be called with lock."""
+        result.finished = True
+        # We need to keep track of the exception so that we can raise it in the same
+        # position every time the iterator is called.
+        result.exception = exception
+        self.generating = False
+        assert self.callable is not None
+        self.generator = self.callable()  # Reset the iterator for the next call.
+        if self.reset_on_exception:
```

Comment:
else we should reset the callable to allow it to be GCed

Reply:
done